### PR TITLE
Promote VisibilityOnDemand to Beta.

### LINKED ISSUE #2973: Promote VisibilityOnDemand to Beta
**What would you like to be added**:

Promote the VisibilityOnDemand to Beta.

**Why is this needed**:

1. the feature is assumed to be enabled by kueuectl "POSITION IN QUEUE" column in the `kueuectl list workload` command
2. The feature is in alpha quite long already
3. The enablement involves 2 steps which is sometimes confusing to users

**Completion requirements**:

This enhancement requires the following artifacts:

- [ ] Design doc
- [ ] API change
- [ ] Docs update

The artifacts should be linked in subsequent comments.

### PR DESCRIPTION
**Pull request by mbobrovskyi**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Promote VisibilityOnDemand to Beta and enabled by default.

#### Which issue(s) this PR fixes:

Fixes #2973

#### Does this PR introduce a user-facing change?

```release-note
Promoted VisibilityOnDemand to Beta and enabled by default.

ACTION REQUIRED: The v1alpha1 Visibility API is deprecated and will be removed in the next release. Please use v1beta1 instead.
```

### CODE REVIEW

**Comment by mimowo** [comment]
/hold
Let's investigate the discussion points first: https://github.com/kubernetes-sigs/kueue/issues/2973#issuecomment-2334286893
and record the findings in the KEP update.

**Comment by alculquicondor** [comment]
I think we can keep the `v1alpha1` API for one release, and put an `ACTION REQUIRED` in the release notes.

File: pkg/visibility/api/install.go [hunk 1]
```diff
@@ -20,42 +22,32 @@ import (
...
-// Install installs API scheme defined in apis/v1alpha1 and registers storage
+// Install installs API scheme and registers storages
 func Install(server *genericapiserver.GenericAPIServer, kueueMgr *queue.Manager) error {
-	apiGroupInfo := genericapiserver.NewDefaultAPIGroupInfo(v1alpha1.GroupVersion.Group, Scheme, ParameterCodec, Codecs)
-	pendingWorkloadsInCqREST := apirest.NewPendingWorkloadsInCqREST(kueueMgr)
-	cqREST := apirest.NewCqREST()
-	pendingWorkloadsInLqREST := apirest.NewPendingWorkloadsInLqREST(kueueMgr)
-	lqREST := apirest.NewLqREST()
-
-	visibilityServerResources := map[string]rest.Storage{
```
Comment (mimowo) [inline pkg/visibility/api/install.go:53]: What happens to this?

File: pkg/visibility/api/rest/v1alpha1/test_utils.go [hunk 2]
```diff
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package rest
+package v1alpha1
```
Comment (mimowo) [inline pkg/visibility/api/rest/v1alpha1/test_utils.go:15]: This seems having nothing to do with the graduation, just a cleanup. Please send a separate PR for the cleanup.

File: site/content/en/docs/installation/_index.md [hunk 3]
```diff
@@ -243,22 +243,23 @@ spec:
 
 The currently supported features are:
 
-| Feature | Default | Stage | Since | Until |
```
Comment (mimowo) [inline site/content/en/docs/installation/_index.md:246]: Could you please reformat the table in a dedicated PR? So that it is clear what is changed.

File: Makefile [hunk 4]
```diff
@@ -265,7 +265,6 @@ artifacts: kustomize yq helm ## Generate release artifacts.
 	$(KUSTOMIZE) build config/dev -o artifacts/manifests-dev.yaml
 	$(KUSTOMIZE) build config/alpha-enabled -o artifacts/manifests-alpha-enabled.yaml
 	$(KUSTOMIZE) build config/prometheus -o artifacts/prometheus.yaml
-	$(KUSTOMIZE) build config/visibility -o artifacts/visibility-api.yaml
```
Comment (mimowo) [inline Makefile:268]: For the ease of reviewing I would suggest to merge-in the yaml in a dedicated PR. WDYT @tenzen-y ?

File: pkg/visibility/api/rest/v1alpha1/test_utils.go (hunk 2 above)
Comment (mbobrovskyi) [reply pkg/visibility/api/rest/v1alpha1/test_utils.go:15]: We need to keep both `v1alpha1` and `v1beta1`. To make it more clear I created two packages for each one.

File: pkg/visibility/api/install.go (hunk 1 above)
Comment (mbobrovskyi) [reply pkg/visibility/api/install.go:53]: The same situation here. We need to keep both `v1alpha1` and `v1beta1` and to optimize code I moved it to separate files.

File: pkg/visibility/api/rest/v1alpha1/test_utils.go (hunk 2 above)
Comment (mbobrovskyi) [reply pkg/visibility/api/rest/v1alpha1/test_utils.go:15]: https://github.com/kubernetes-sigs/kueue/blob/b765d06caeae4c380f7f9555092e960ac6e31450/pkg/visibility/api/rest/v1alpha1/storage.go#L25-L32

File: pkg/visibility/api/install.go (hunk 1 above)
Comment (mbobrovskyi) [reply pkg/visibility/api/install.go:53]: https://github.com/kubernetes-sigs/kueue/blob/b765d06caeae4c380f7f9555092e960ac6e31450/pkg/visibility/api/rest/v1alpha1/storage.go#L25-L32

File: site/content/en/docs/installation/_index.md (hunk 3 above)
Comment (mbobrovskyi) [reply site/content/en/docs/installation/_index.md]: Done

File: pkg/visibility/api/rest/v1alpha1/test_utils.go (hunk 2 above)
Comment (mimowo) [reply pkg/visibility/api/rest/v1alpha1/test_utils.go:15]: Right, but still this change (rename rest -> v1alpha1) could be done as a preparatory PR before the main PR?

File: pkg/visibility/api/install.go (hunk 1 above)
Comment (mimowo) [reply pkg/visibility/api/install.go:53]: I see, sgtm, we may just call it `NewStorage`.

File: pkg/visibility/api/rest/v1alpha1/test_utils.go (hunk 2 above)
Comment (mbobrovskyi) [reply pkg/visibility/api/rest/v1alpha1/test_utils.go:15]: I can prepare another one.

Comment (mbobrovskyi) [reply pkg/visibility/api/rest/v1alpha1/test_utils.go:15]: Done on https://github.com/kubernetes-sigs/kueue/pull/3041.

File: site/content/en/docs/installation/_index.md (hunk 3 above)
Comment (mbobrovskyi) [reply site/content/en/docs/installation/_index.md]: Done on https://github.com/kubernetes-sigs/kueue/pull/3048.

File: Makefile (hunk 4 above)
Comment (tenzen-y) [reply Makefile:268]: Yeah, we can move manifests in a dedicated PR.

Comment (mbobrovskyi) [reply Makefile:268]: Created separate PR for this changes https://github.com/kubernetes-sigs/kueue/pull/3084.

**Comment by mbobrovskyi** [comment]
@mimowo Could you please take another look?

File: site/content/en/docs/tasks/manage/monitor_pending_workloads/pending_workloads_on_demand.md [hunk 5]
```diff
@@ -260,15 +262,15 @@ You can pass optional query parameters:
 To view only 1 pending workloads use, starting from position 1 in LocalQueue run:
 
 ```shell
-kubectl get --raw "/apis/visibility.kueue.x-k8s.io/v1alpha1/localqueues/user-queue/pendingworkloads?limit=1&offset=1"
+kubectl get --raw "/apis/visibility.kueue.x-k8s.io/v1beta1/localqueues/user-queue/pendingworkloads?limit=1&offset=1"
```
Comment (mbobrovskyi) [inline site/content/en/docs/tasks/manage/monitor_pending_workloads/pending_workloads_on_demand.md:265]:
```suggestion
kubectl get --raw "/apis/visibility.kueue.x-k8s.io/v1alpha1/namespaces/default/localqueues/user-queue/pendingworkloads?limit=1&offset=1"
```

Looks like here not correct command. Created separate PR https://github.com/kubernetes-sigs/kueue/pull/3097 to fix on website.

Comment (mbobrovskyi) [reply site/content/en/docs/tasks/manage/monitor_pending_workloads/pending_workloads_on_demand.md]: Done

**Comment by PBundyra** [comment]
This looks great, thank you!
/lgtm

**Review by mimowo — COMMENTED** [review]
/approve 
Thanks for everyone involved!

**Comment by mbobrovskyi** [comment]
/unhold 

Due to P&F configs added on https://github.com/kubernetes-sigs/kueue/pull/3043.
